graphql/util.go

```diff
@@ -1290,3 +1290,67 @@ func getAPISubscriptionsForProject(ctx context.Context, projectId string) ([]*re
 	}
 	return res, nil
 }
+
+// GroupProjects takes a list of projects and groupes them by their repo. If onlyDefaultedToRepo is true,
+// it only groups projects that are defaulted to the repo
+func GroupProjects(projects []model.ProjectRef, onlyDefaultedToRepo bool) ([]*GroupedProjects, error) {
+	groupsMap := make(map[string][]*restModel.APIProjectRef)
+
+	for _, p := range projects {
+		groupName := fmt.Sprintf("%s/%s", p.Owner, p.Repo)
+
+		if onlyDefaultedToRepo && p.UseRepoSettings {
+			groupName = ""
+		}
+
+		apiProjectRef := restModel.APIProjectRef{}
+		if err := apiProjectRef.BuildFromService(p); err != nil {
+			return nil, errors.Wrap(err, "error building APIProjectRef from service")
+		}
+
+		if projs, ok := groupsMap[groupName]; ok {
+			groupsMap[groupName] = append(projs, &apiProjectRef)
+		} else {
+			groupsMap[groupName] = []*restModel.APIProjectRef{&apiProjectRef}
+		}
+	}
+
+	groupsArr := []*GroupedProjects{}
+
+	for groupName, groupedProjects := range groupsMap {
+		gp := GroupedProjects{
+			Name:             &groupName, //deprecated
+			GroupDisplayName: groupName,
+			Projects:         groupedProjects,
+		}
+
+		if groupedProjects[0].UseRepoSettings {
+			repoRef, err := model.FindOneRepoRef(utility.FromStringPtr(groupedProjects[0].RepoRefId))
+			if err != nil {
+				return nil, err
+			}
+
+			apiRepoRef := restModel.APIProjectRef{}
+			if repoRef != nil {
+				if err := apiRepoRef.BuildFromService(repoRef.ProjectRef); err != nil {
+					return nil, errors.Wrap(err, "error building the repo's ProjectRef from service")
+				}
+			}
+			if repoRef != nil {
+				gp.Repo = &apiRepoRef
+				if repoRef.DisplayName != "" {
```

Comment:
Can you explain what this is doing?
This looks pretty expensive if we have to do this query for every project we are returning. (There appears to be > 100 in the dropdown)

Reply:
This will only be fetched for projects that default to the repo. I suspect that there won't be too many, especially to start. It fetches the repoRef so that we can access the repoId (so that we can use it for routing, so that the repo itself is clickable), and so that we can use the repoDisplayName if the user set it instead of owner/repo.

Reply:
It's also not for every project, it's for every owner/repo pair.